### PR TITLE
Bug fix about receive endline

### PR DESCRIPTION
**Pull request by heat1024**
Sometimes, in server, there is a lot of CLOSE_WAIT connections between pftp and origin ftp server because some client send wrong commands.

These wrong commands makes origin ftp server stuck in hang status.

The commands just like this.
```
[root@syslog001 ~]# zcat /var/log/dyconftp/dyconftp.log_201904*.gz | grep -F '10.51.80.96:37998'
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=info msg="FTP Client connected clientIp 10.51.80.96:37998"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=debug msg="[91] addr:10.51.80.96:37998 new proxy from=127.0.0.1:38140 to=127.0.0.1:21"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=info msg="[91] addr:10.51.80.96:37998 read from client: PROXY TCP4 114.178.135.122 133.130.34.170 61621 21\r\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=debug msg="[91] addr:10.51.80.96:37998 response from origin: 220 Lolipop FTP server ready\r\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=info msg="[91] addr:10.51.80.96:37998 read from client: \x16\x03\x01\x01<\x01\x00\x018\x03\x03\\\xa1\x82\x87\xc7\xfd0\xa4/\x94\x18\xea\xa9@\x1bq\x96hʂ\\:\xb0\x9a0\x93ϯ\xea\xf4\xb5[\x00\x00\xa0\xc00\xc0,\xc0(\xc0$\xc0\x14\xc0\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=debug msg="[91] addr:10.51.80.96:37998 send to origin: \x16\x03\x01\x01<\x01\x00\x018\x03\x03\\\xa1\x82\x87\xc7\xfd0\xa4/\x94\x18\xea\xa9@\x1bq\x96hʂ\\:\xb0\x9a0\x93ϯ\xea\xf4\xb5[\x00\x00\xa0\xc00\xc0,\xc0(\xc0$\xc0\x14\xc0\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=info msg="[91] addr:10.51.80.96:37998 read from client: \xc0\"\xc0!\x00\xa3\x00\x9f\x00k\x00j\x009\x008\x00\x88\x00\x87\xc02\xc0.\xc0*\xc0&\xc0\x0f\xc0\x05\x00\x9d\x00=\x005\x00\x84\xc0\x12\xc0\b\xc0\x1c\xc0\x1b\x00\x16\x00\x13\xc0\r\xc0\x03\x00\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=debug msg="[91] addr:10.51.80.96:37998 send to origin: \xc0\"\xc0!\x00\xa3\x00\x9f\x00k\x00j\x009\x008\x00\x88\x00\x87\xc02\xc0.\xc0*\xc0&\xc0\x0f\xc0\x05\x00\x9d\x00=\x005\x00\x84\xc0\x12\xc0\b\xc0\x1c\xc0\x1b\x00\x16\x00\x13\xc0\r\xc0\x03\x00\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=info msg="[91] addr:10.51.80.96:37998 read from client: \xc0/\xc0+\xc0'\xc0#\xc0\x13\xc0\t\xc0\x1f\xc0\x1e\x00\xa2\x00\x9e\x00g\x00@\x003\x002\x00\x9a\x00\x99\x00E\x00D\xc01\xc0-\xc0)\xc0%\xc0\x0e\xc0\x04\x00\x9c\x00<\x00/\x00\x96\x00A\x00\a\xc0\x11\xc0\a\xc0\f\xc0\x02\x00\x05\x00\x04\x00\x15\x00\x12\x00\t\x00\x14\x00\x11\x00\b\x00\x06\x00\x03\x00\xff\x01\x00\x00o\x00\v\x00\x04\x03\x00\x01\x02\x00\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=debug msg="[91] addr:10.51.80.96:37998 send to origin: \xc0/\xc0+\xc0'\xc0#\xc0\x13\xc0\t\xc0\x1f\xc0\x1e\x00\xa2\x00\x9e\x00g\x00@\x003\x002\x00\x9a\x00\x99\x00E\x00D\xc01\xc0-\xc0)\xc0%\xc0\x0e\xc0\x04\x00\x9c\x00<\x00/\x00\x96\x00A\x00\a\xc0\x11\xc0\a\xc0\f\xc0\x02\x00\x05\x00\x04\x00\x15\x00\x12\x00\t\x00\x14\x00\x11\x00\b\x00\x06\x00\x03\x00\xff\x01\x00\x00o\x00\v\x00\x04\x03\x00\x01\x02\x00\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=info msg="[91] addr:10.51.80.96:37998 read from client: \x004\x002\x00\x0e\x00\r\x00\x19\x00\v\x00\f\x00\x18\x00\t\x00\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=debug msg="[91] addr:10.51.80.96:37998 send to origin: \x004\x002\x00\x0e\x00\r\x00\x19\x00\v\x00\f\x00\x18\x00\t\x00\n"
Apr  1 12:16:21 dyconftp-1 dyconftp[5830]: time="2019-04-01T12:16:21+09:00" level=info msg="[91] addr:10.51.80.96:37998 client disconnected by timeout"
```

I had reappearance this situation in my local environment with docker.

This PR make block the wrong command line from client what has not end with `\r\n`.
If client send wrong command, pftp doesn't send command to origin and return error to client.

+ connection to origin server disconnect immediately when connection between client got error.

### CODE REVIEW

File: pftp/proxy.go [hunk 1]
```diff
@@ -83,6 +83,10 @@ func newProxyServer(conf *proxyServerConfig) (*proxyServer, error) {
 }
 
 func (s *proxyServer) sendToOrigin(line string) error {
+	if !strings.HasSuffix(line, "\r\n") {
```
Comment (pyama86) [inline pftp/proxy.go]: Is it okay to add `"\r\n"`?

Comment (heat1024) [reply pftp/proxy.go]: Basically, the "normal" FTP clients(even telnet client) always send command line with CRLF.

The blocking works just for abnormal messages from client. (like some kind of hardware control packets and the others)

Of course if pftp fix wrong commands, ftp server can handle the command.

So, I accept your opinion and fixed the PR codes!

https://github.com/pyama86/pftp/pull/29/commits/1159fc8f509fd405a541c418ecf612a49ad29512

Comment (pyama86) [reply pftp/proxy.go]: i understand.
thanks your commit.

**Comment by heat1024** [comment]
CI-test failed continuously.

In CI-test log, `read from xxx` or `response from xxx` are not shown.
(except response from origin: 211-Features:)

And may be response from pftp -> client port is not working.

I want figure out what makes problem like this.
Is there any changes about some kind of travis CI setting?

```
INFO[0063] FTP Client connected clientIp 127.0.0.1:45806 
DEBU[0063] [6] addr:127.0.0.1:45806 new proxy from=127.0.0.1:56944 to=127.0.0.1:21 
 
INFO[0063] FTP Client connected clientIp 127.0.0.1:45807 
 
 
DEBU[0063] [7] addr:127.0.0.1:45807 new proxy from=127.0.0.1:56948 to=127.0.0.1:21 
DEBU[0063] [6] addr:127.0.0.1:45806 response from origin: 211-Features:
 CCC
 SITE MKDIR
 PBSZ
 AUTH TLS
 MFF modify;UNIX.group;UNIX.mode;
 REST STREAM
 MLST modify*;perm*;size*;type*;unique*;UNIX.group*;UNIX.mode*;UNIX.owner*;
 UTF8
 LANG en-US*
 EPRT
 SITE SYMLINK
 EPSV
 SITE UTIME
 MDTM
 SITE RMDIR
 SSCN
 TVFS
 SITE COPY
 MFMT
 SIZE
 PROT
 
 
DEBU[0063] [6] addr:127.0.0.1:45806 suspend proxy       
INFO[0063] [6] addr:127.0.0.1:45806 switch origin to: 127.0.0.1:10021 
 
INFO[0063] FTP Client connected clientIp 127.0.0.1:45810 
 
 
DEBU[0063] [6] addr:127.0.0.1:45806 current connection count: 1 
DEBU[0063] [6] addr:127.0.0.1:45806 unsuspend proxy     
ERRO[0063] [6] addr:127.0.0.1:45806 command error response: read tcp 127.0.0.1:2121->127.0.0.1:45806: i/o timeout 
DEBU[0063] [6] addr:127.0.0.1:45806 send to client:421 command timeout : closing control connection 
INFO[0063] [6] addr:127.0.0.1:45806 client disconnected by timeout 
ERRO[0063] read tcp 127.0.0.1:2121->127.0.0.1:45806: i/o timeout 
DEBU[0063] [8] addr:127.0.0.1:45810 new proxy from=127.0.0.1:56956 to=127.0.0.1:21 
```

**Comment by pyama86** [comment]
I have not changed anything.
Does `make ci` complete successfully in your laptop?

**Comment by heat1024** [comment]
Yes. In my laptop, it works perfectly

**Comment by pyama86** [comment]
realy?
I will also check.
please wait.

**Comment by heat1024** [comment]
This is the failing sequence what I think.
1. the integration client connect to ftp
2. client send FEAT to origin
3. client disconnected from pftp by abnormal
4. origin response about FEAT
5. pftp send to response to client, but failed

**Comment by pyama86** [comment]
Tests will also fall on the master branch...

**Comment by heat1024** [comment]
Oh...
Then it is sure something's changed.

**Comment by heat1024** [comment]
The installed go mod package versions are same with Laptop.

**Comment by heat1024** [comment]
Oh... I found the problem.

It was vsftpd image.

vsftpd's latest docker image makes error like this

```
# in integration test
DEBU[0017] [12] addr:127.0.0.1:49838 response from new origin: 500 OOPS: failed to open xferlog log file:/var/log/vsftpd/vsftpd.log
DEBU[0017] [12] addr:127.0.0.1:49838 send to origin: USER vsuser
DEBU[0017] [12] addr:127.0.0.1:49838 current connection count: 2 
DEBU[0017] [12] addr:127.0.0.1:49838 unsuspend proxy    
ERRO[0017] [12] addr:127.0.0.1:49838 command error response: read tcp 127.0.0.1:2121->127.0.0.1:49838: i/o timeout 

# when connect vsftpd directly
❯ lftp -u vsuser,vsuser 127.0.0.1 -p 10021 -e 'set ftp:ssl-allow no; set ssl:verify-certificate no; set ftp:passive-mode yes; set ftp:prefer-epsv no'
lftp vsuser@127.0.0.1:~> debug 10
lftp vsuser@127.0.0.1:~> ls
---- dns cache hit
---- attempt number 1 (max_retries=1000)
---- Connecting to 127.0.0.1 (127.0.0.1) port 10021
<--- 500 OOPS: failed to open xferlog log file:/var/log/vsftpd/vsftpd.log
---> FEAT
buffer: EOF on FD 4
---- Closing control socket
Interrupt                                                                           
lftp vsuser@127.0.0.1:~> quit
```

So vsftpd tests are failed.


I'll fix it!

**Comment by heat1024** [comment]
Error might appeared after this PR...
https://github.com/fauria/docker-vsftpd/commit/8b415e968e79c1464aeecb21908e741670b252fc

It duplicated with previous dockerfile setup.
So I delete that line.

```Diff
FROM fauria/vsftpd
ENV FTP_USER="vsuser"
ENV FTP_PASS="vsuser"
ENV PASV_ADDRESS="127.0.0.1"
ENV PASV_MIN_PORT="11100"
ENV PASV_MAX_PORT="11110"
-ENV LOG_STDOUT=""
ENV LOCAL_UMASK="022"
VOLUME /home/vsftpd
VOLUME /var/log/vsftpd
RUN echo "log_ftp_protocol=YES" >> /etc/vsftpd/vsftpd.conf
EXPOSE 20 21

CMD ["/usr/sbin/run-vsftpd.sh"]

```

**Comment by pyama86** [comment]
yeah!lgtml!

**Comment by heat1024** [comment]
Thanks for review!

**Pull request merged**